File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def helper(i,subset):
            if len(subset)==k:
                res.append(subset[:])
                return
            if i>n:
                return
            subset.append(i)
            helper(i+1,subset)
            subset.pop()
            helper(i+1,subset)
        #---------------------
        res=[]
        helper(1,[])
        return res

File: 77-combinations/combinations.java
class Solution {
    public void backtracking(int n,List<List<Integer>> res, List<Integer> subset,int i,int k){
        if(subset.size()==k){
                res.add(new ArrayList<>(subset));
        }
        for(int j=i;j<=n;j++){
            subset.add(j);
            backtracking(n,res,subset,j+1,k);
            subset.remove(subset.size()-1);
        }
    }
    public List<List<Integer>> combine(int n, int k)  {
        List<List<Integer>> res=new ArrayList<>();
        List<Integer> subset=new ArrayList<>();
        backtracking(n,res,subset,1,k); // subset=[]  i=0
        return res;
    }
}